SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift

```diff
@@ -0,0 +1,264 @@
+//===--- LoadCopyToBorrowOptimization.swift --------------------------------==//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Replaces a `load [copy]` with a `load_borrow` if possible.
+///
+/// ```
+///   %1 = load [copy] %0
+///   // no writes to %0
+///   destroy_value %1
+/// ```
+/// ->
+/// ```
+///   %1 = load_borrow %0
+///   // no writes to %0
+///   end_borrow %1
+/// ```
+///
+/// The optimization can be done if:
+/// * During the (forward-extended) lifetime of the loaded value the memory location is not changed.
+/// * All (forward-extended) uses of the loaded value support guaranteed ownership.
+/// * The (forward-extended) lifetime of the loaded value ends with `destroy_value`(s).
+///
+let loadCopyToBorrowOptimization = FunctionPass(name: "load-copy-to-borrow-optimization") {
+  (function: Function, context: FunctionPassContext) in
+
+  if !function.hasOwnership {
+    return
+  }
+
+  for inst in function.instructions {
+    if let load = inst as? LoadInst {
+      optimize(load: load, context)
+    }
+  }
+}
+
+private func optimize(load: LoadInst, _ context: FunctionPassContext) {
+  if load.loadOwnership != .copy {
+    return
+  }
+
+  var ownershipUses = OwnershipUses(context)
+  defer { ownershipUses.deinitialize() }
+
+  // Handle the special case if the `load` is immediately followed by a single `move_value`.
+  // Note that `move_value` is _not_ a forwarding instruction.
+  // In this case we have to preserve the move's flags by inserting a `begin_borrow` with the same flags.
```

Comment:
Would this portion of the comment be most useful in `replaceWithLoadBorrow(liferange:ownershipUses:)` near
```
     if let moveInst = singleMoveValueUser {
```
?